gradio/processing_utils.py

```diff
@@ -518,3 +522,61 @@ def _scale(a, n, m, copy=True):
 
 def strip_invalid_filename_characters(filename: str) -> str:
     return "".join([char for char in filename if char.isalnum() or char in "._- "])
+
+
+def ffmpeg_installed() -> bool:
+    return shutil.which("ffmpeg") is not None
+
+
+def video_is_playable(video_filepath: str) -> bool:
+    """Determines if a video is playable in the browser.
+
+    A video is playable if it has a playable container and codec.
+        .mp4 -> h264
+        .webm -> vp9
+        .ogg -> theora
+    """
+    try:
+        container = pathlib.Path(video_filepath).suffix.lower()
+        output = subprocess.run(
+            [
+                "ffprobe",
+                "-show_format",
+                "-show_streams",
+                "-select_streams",
+                "v",
+                "-print_format",
+                "json",
+                video_filepath,
+            ],
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+        )
+        output = json.loads(output.stdout)
+        video_codec = output["streams"][0]["codec_name"]
+        return (container, video_codec) in [
+            (".mp4", "h264"),
+            (".ogg", "theora"),
+            (".webm", "vp9"),
+        ]
+    # If anything goes wrong, assume the video can be played to not convert downstream
+    except Exception:
+        return True
+
+
+def convert_video_to_playable_mp4(video_path: str) -> str:
+    """Convert the video to mp4. If something goes wrong return the original video."""
+    try:
+        output_path = pathlib.Path(video_path).with_suffix(".mp4")
```

Comment:
I don't understand why we create this `output_path` variable and then copy the original file to it. 

1. Shouldn't we just pass as the input to `FFmpeg` the original `video_path`?
2. If there's an exception in `FFmpeg`, it looks like we would return the original video file but with the ".mp4" suffix. Wouldn't this cause issues if the video is in a different format?

Reply:
1. The converted video needs to have an extension that's playable in the browser (`mp4`, `ogg`, `webm`). If not, the browser won't play it (even if it has a playable video codec) so this function has to rename the extension (I arbitrarily picked mp4).
2. If there's an exception, I think we'll return the original video which is what's tested with this unit [test](https://github.com/gradio-app/gradio/pull/2003/files#diff-c7fbfc627571585283ab24f28191f1e640428c3d91159faf0c3e393ab7759f03R164)

Reply:
Oh sorry I misread the code, yeah this is completely fine!